src/ipc/ipc4/handler.c

```diff
@@ -965,6 +965,37 @@ void ipc_msg_reply(struct sof_ipc_reply *reply)
 	ipc_compound_msg_done(in.primary.r.type, reply->error);
 }
 
+void ipc_msg_send_status(void)
+{
+	char *data = ipc_get()->comp_data;
+	int count = 10;
+	int ret;
+
+	/* copy mailbox data to message */
+	if (msg_reply.tx_size > 0 && msg_reply.tx_size < SOF_IPC_MSG_MAX_SIZE) {
+		ret = memcpy_s(msg_reply.tx_data, msg_reply.tx_size, data, msg_reply.tx_size);
+		if (ret) {
+			tr_err(&ipc_tr, "failed to copy memory %d", ret);
+			return;
+		}
+	}
+
+	/* send critical status to host since host is waiting for it */
+	while (count) {
+		ret = ipc_platform_send_msg(&msg_reply);
```

Comment:
hrm, is this looping needed?

Reply:
wait for ipc hw idle since FW can send notification msg to host and this may compete with reply msg

Reply:
can this loop lead to missing a host message? You send the reply, for some reason you get an error back, but the actual reply is delivered to the host. Then the host sends the next IPC and you reply to it here again - without processing it

Reply:
@lyakh, I believe the implementation here relies on the fact that src/drivers/intel/cavs/ipc.c : ipc_platform_send_msg() will only return error if there is an IPC from DSP to host is pending (sent, but not cleared by the host)?
In IPC4 the reply and the notification are sent a same way: as a FW initiated IPC message.
The host is responsible to serialize it's IPC tx, which is send message, get reply (and an ack in between).

@RanderWang, but if you set the `high_priority` to true, then ipc_msg_send() will first try to send the message asap, if we are busy then the message will be added to the head of the msg_list. When the message which blocked the immediate sending is done then we will pick a message from the list right away and send it, no?

Reply:
please check the following function. queued msg is sent when there is a interrupt coming, or fw will sleep and the queued msg can't be sent.  IPC3 doesn't have such issue as IPC4

````c
enum task_state task_main_primary_core(void *data)
{
	struct ipc *ipc = ipc_get();

	/* main audio processing loop */
	while (1) {
		/* sleep until next IPC or DMA */
		wait_for_interrupt(0);

		if (!ipc->pm_prepare_D3)
			ipc_send_queued_msg();

	}

	return SOF_TASK_STATE_COMPLETED;
}
````